counsel.el

```diff
@@ -191,11 +191,9 @@ Update the minibuffer with the amount of lines collected every
       (with-current-buffer (process-buffer process)
         (goto-char (point-min))
         (setq size (- (buffer-size) (forward-line (buffer-size))))
-        (ivy--set-candidates
-         (split-string
-          (buffer-string)
-          counsel-async-split-string-re
-          t)))
+        (let ((cands (split-string (buffer-string) counsel-async-split-string-re t)))
+          (setq ivy--old-cands ivy--all-candidates)
```

Comment:
Ok, I finally had some time to look at this a little harder.

Lets forget about what I was saying about the killed bit, it turns out that the fact that the process is killed and not handled in `counsel--async-sentinel` is not relevant to the change I'm trying to make or the issue I was seeing.

Originally, when I factored out the common code into `counsel-grep-mode-occur`, I was doing an `ivy--filter` over `ivy--all-candidates` before filling the occur buffer.  However, I noticed that if I interrupted a running `counsel-ag` (e.g. say you look for "TCP" in the linux kernel source and `C-c C-o` after the first few matches, since running that generally takes a while), then the occur buffer would have no candidates, even though the mini buffer had shown several.

Now, before I actually submitted the pull request, I found myself wondering why I needed to use `ivy--filter` at all, since I thought all the functions I was factoring out would do the filtering themselves.  So, I removed the `ivy--filter` call from `counsel-grep-mode-occur`, but I accidentally left this line in.  That was just a mistake.

But, as you pointed out, not using `ivy--filter` broke `counsel-git-grep-occur` for repositories less than 20k lines, so now I'm back to the original issue.

What winds up happening is that `ivy--filter` is called when `ivy--old-re` is set (e.g. its set to "TCP" in the example I gave as `counsel-ag-function` sets it before running `ag`), but `ivy--old-cands` is nil.  This triggers the second clause of the `cond` statement in `ivy--filter`.  So, I was trying to fix the issue by accumulating entries into `ivy--old-cands` as the `ag` command was running such that when that clause of `ivy--filter` was hit, it would return the correct results. 

But, you could also remove this line and instead add a condition to that `cond` that `ivy--old-cands` be non nil (e.g. `(and ivy--old-re ivy--old-cands (stringp re) ...`).  I've tested this and this seems to work.  But, to be honest, I've stared at that second conditional case for a bit and I'm honestly not quite sure what is purpose is, so I'm not sure if doing what I just described actually makes sense.

Do you have suggestions about the appropriate way to address the issue I've described?